File: roboplan/test/test_scene.cpp
#include <gmock/gmock.h>
#include <gtest/gtest.h>
#include <memory>
#include <vector>

#include <roboplan/core/path_utils.hpp>
#include <roboplan/core/scene.hpp>
#include <roboplan_example_models/resources.hpp>

namespace {
constexpr double kTolerance = 1e-4;
}

namespace roboplan {

using ::testing::ContainerEq;
using ::testing::Not;

class RoboPlanSceneTest : public ::testing::Test {
protected:
  void SetUp() override {
    const auto share_prefix = example_models::get_install_prefix() / "share";
    const auto model_prefix = share_prefix / "roboplan_example_models" / "models";
    urdf_path_ = model_prefix / "ur_robot_model" / "ur5_gripper.urdf";
    srdf_path_ = model_prefix / "ur_robot_model" / "ur5_gripper.srdf";
    package_paths_ = {share_prefix};
    yaml_config_path_ = model_prefix / "ur_robot_model" / "ur5_config.yaml";
    scene_ = std::make_unique<Scene>("test_scene", urdf_path_, srdf_path_, package_paths_,
                                     yaml_config_path_);
  }

public:
  // No default constructor, so must be a pointer.
  std::unique_ptr<Scene> scene_;
  std::filesystem::path urdf_path_;
  std::filesystem::path srdf_path_;
  std::vector<std::filesystem::path> package_paths_;
  std::filesystem::path yaml_config_path_;
};

TEST_F(RoboPlanSceneTest, SceneProperties) {
  EXPECT_EQ(scene_->getName(), "test_scene");
  EXPECT_EQ(scene_->getModel().nq, 6u);
  EXPECT_THAT(scene_->getJointNames(),
              ContainerEq(std::vector<std::string>{"shoulder_pan_joint", "shoulder_lift_joint",
                                                   "elbow_joint", "wrist_1_joint", "wrist_2_joint",
                                                   "wrist_3_joint"}));

  const auto maybe_joint_info = scene_->getJointInfo("shoulder_pan_joint");
  ASSERT_TRUE(maybe_joint_info.has_value()) << maybe_joint_info.error();
  const auto& joint_info = maybe_joint_info.value();
  EXPECT_EQ(joint_info.type, JointType::REVOLUTE);
  EXPECT_EQ(joint_info.num_position_dofs, 1u);
  EXPECT_EQ(joint_info.num_velocity_dofs, 1u);
  ASSERT_EQ(joint_info.limits.min_position.size(), 1u);
  EXPECT_NEAR(joint_info.limits.min_position[0], -M_PI, kTolerance);
  ASSERT_EQ(joint_info.limits.max_position.size(), 1u);
  EXPECT_NEAR(joint_info.limits.max_position[0], M_PI, kTolerance);
  ASSERT_EQ(joint_info.limits.max_velocity.size(), 1u);
  EXPECT_NEAR(joint_info.limits.max_velocity[0], 3.15, kTolerance);
  ASSERT_EQ(joint_info.limits.max_acceleration.size(), 1u);
  EXPECT_NEAR(joint_info.limits.max_acceleration[0], 2.0, kTolerance);
  ASSERT_EQ(joint_info.limits.max_jerk.size(), 1u);
  EXPECT_NEAR(joint_info.limits.max_jerk[0], 10.0, kTolerance);

  std::cout << *scene_;  // Test printing for good measure
}

TEST_F(RoboPlanSceneTest, RandomPositions) {
  // Test subsequent pseudorandom values.
  const auto orig_random_positions = scene_->randomPositions();
  const auto new_random_positions = scene_->randomPositions();
  EXPECT_EQ(orig_random_positions.size(), 6);
  EXPECT_THAT(orig_random_positions, Not(ContainerEq(new_random_positions)));

  // Test seeded values.
  scene_->setRngSeed(1234);
  const auto orig_seeded_positions = scene_->randomPositions();
  EXPECT_EQ(orig_seeded_positions.size(), 6);
  scene_->setRngSeed(1234);  // reset seed
  const auto new_seeded_positions = scene_->randomPositions();
  EXPECT_THAT(orig_seeded_positions, ContainerEq(new_seeded_positions));
}

TEST_F(RoboPlanSceneTest, CollisionCheck) {
  // Collision free
  Eigen::VectorXd q_free(6);
  q_free << 0.0, -1.57, 0.0, 0.0, 0.0, 0.0;
  EXPECT_FALSE(scene_->hasCollisions(q_free));

  // In collision
  Eigen::VectorXd q_coll(6);
  q_coll << 0.0, -1.57, 3.0, 0.0, 0.0, 0.0;
  EXPECT_TRUE(scene_->hasCollisions(q_coll));
}

TEST_F(RoboPlanSceneTest, CollisionCheckAlongPath) {
  // Collision free
  Eigen::VectorXd q_start_free(6);
  q_start_free << 0.0, -1.57, 0.0, 0.0, 0.0, 0.0;
  Eigen::VectorXd q_end_free(6);
  q_end_free << 1.0, -1.57, 1.57, 0.0, 0.0, 0.0;
  EXPECT_FALSE(hasCollisionsAlongPath(*scene_, q_start_free, q_end_free, 0.05));

  Eigen::VectorXd q_end_coll(6);
  q_end_coll << 0.0, -1.57, 3.0, 0.0, 0.0, 0.0;
  EXPECT_TRUE(hasCollisionsAlongPath(*scene_, q_start_free, q_end_coll, 0.05));
}

TEST_F(RoboPlanSceneTest, GetFrameMapReturnsCorrectMapping) {
  const auto model = scene_->getModel();

  // Verify the frame IDs are correct
  for (const auto& frame : model.frames) {
    if (frame.name == "universe")
      continue;
    EXPECT_EQ(scene_->getFrameId(frame.name), model.getFrameId(frame.name));
  }
}

TEST_F(RoboPlanSceneTest, TestForwardKinematics) {
  // Collision free
  Eigen::VectorXd q(6);
  q << 0.0, 0.0, 0.0, 0.0, 0.0, 0.0;
  const auto fk = scene_->forwardKinematics(q, "tool0");

  // Expected transformation matrix for zero configuration for a UR5e
  Eigen::Matrix4d expected;
  expected << -1.0, 0.0, 0.0, 0.81725, 0.0, 0.0, 1.0, 0.19145, 0.0, 1.0, 0.0, -0.005491, 0.0, 0.0,
      0.0, 1.0;

  EXPECT_TRUE(fk.isApprox(expected, 1e-6));
}

TEST_F(RoboPlanSceneTest, TestLoadXMLStrings) {
  // Load the sample XMLs from file as strings.
  auto urdf_xml = readFile(urdf_path_);
  auto srdf_xml = readFile(srdf_path_);

  // Just make sure it is the same as when loading from file (the validation is above)
  auto scene_xml =
      std::make_unique<Scene>("test_scene", urdf_xml, srdf_xml, package_paths_, yaml_config_path_);
  EXPECT_EQ(scene_xml->getModel().nq, scene_->getModel().nq);
  EXPECT_THAT(scene_xml->getJointNames(), scene_->getJointNames());

  const auto seeded_positions = scene_xml->randomPositions();
  EXPECT_EQ(seeded_positions.size(), 6);
}

TEST_F(RoboPlanSceneTest, TestCollisionGeometry) {
  // Nominally, this configuration is collision free.
  Eigen::VectorXd q(6);
  q << 0.0, 0.0, 0.0, 0.0, 0.0, 0.0;
  ASSERT_FALSE(scene_->hasCollisions(q));

  // Add some collision objects
  const auto color = Eigen::Vector4d(0.5, 0.5, 0.5, 0.5);

  Eigen::Matrix4d box_tform = Eigen::Matrix4d::Identity();
  box_tform(2, 3) = 1.0;  // z position
  const auto add_box_result =
      scene_->addBoxGeometry("test_box", "universe", Box(1.0, 1.0, 0.5), box_tform, color);
  ASSERT_TRUE(add_box_result.has_value()) << add_box_result.error();

  Eigen::Matrix4d sphere_tform = Eigen::Matrix4d::Identity();
  sphere_tform(0, 3) = 3.0;  // x position
  const auto add_sphere_result =
      scene_->addSphereGeometry("test_sphere", "universe", Sphere(0.5), sphere_tform, color);
  ASSERT_TRUE(add_sphere_result.has_value()) << add_sphere_result.error();

  ASSERT_FALSE(scene_->hasCollisions(q));  // should still be collision free

  // Now move one of the collision objects to be in collision.
  sphere_tform(0, 3) = 0.0;
  const auto move_sphere_result =
      scene_->updateGeometryPlacement("test_sphere", "universe", sphere_tform);
  ASSERT_TRUE(move_sphere_result.has_value()) << move_sphere_result.error();
  ASSERT_TRUE(scene_->hasCollisions(q));

  // Remove the collision object and verify that the robot is no longer in collision.
  const auto remove_sphere_result = scene_->removeGeometry("test_sphere");
  ASSERT_TRUE(remove_sphere_result.has_value()) << remove_sphere_result.error();
  ASSERT_FALSE(scene_->hasCollisions(q));
}

}  // namespace roboplan

File: bindings/test/test_scene.py
"""
Unit tests for scenes in RoboPlan.
"""

from pathlib import Path

import pytest
import numpy as np
import pinocchio as pin

from roboplan import (
    get_install_prefix,
    hasCollisionsAlongPath,
    Box,
    JointType,
    Scene,
    Sphere,
)


@pytest.fixture
def test_scene() -> Scene:
    roboplan_examples_dir = Path(get_install_prefix()) / "share"
    roboplan_models_dir = roboplan_examples_dir / "roboplan_example_models" / "models"
    urdf_path = roboplan_models_dir / "ur_robot_model" / "ur5_gripper.urdf"
    srdf_path = roboplan_models_dir / "ur_robot_model" / "ur5_gripper.srdf"
    package_paths = [roboplan_examples_dir]
    yaml_config_path = roboplan_models_dir / "ur_robot_model" / "ur5_config.yaml"

    return Scene("test_scene", urdf_path, srdf_path, package_paths, yaml_config_path)


def test_scene_properties(test_scene: Scene) -> None:
    assert test_scene.getName() == "test_scene"
    assert test_scene.getJointNames() == [
        "shoulder_pan_joint",
        "shoulder_lift_joint",
        "elbow_joint",
        "wrist_1_joint",
        "wrist_2_joint",
        "wrist_3_joint",
    ]

    joint_info = test_scene.getJointInfo("shoulder_pan_joint")
    assert joint_info.type == JointType.REVOLUTE
    assert joint_info.num_position_dofs == 1
    assert joint_info.num_velocity_dofs == 1
    assert np.allclose(joint_info.limits.min_position, np.array([-np.pi]))
    assert np.allclose(joint_info.limits.max_position, np.array([np.pi]))
    assert np.allclose(joint_info.limits.max_velocity, np.array([3.15]))
    assert np.allclose(joint_info.limits.max_acceleration, np.array([2.0]))
    assert np.allclose(joint_info.limits.max_jerk, np.array([10.0]))

    print(test_scene)  # Test printing for good measure


def test_random_positions(test_scene: Scene) -> None:
    # Test subsequent pseudorandom values.
    orig_random_positions = test_scene.randomPositions()
    new_random_positions = test_scene.randomPositions()
    assert np.all(np.not_equal(orig_random_positions, new_random_positions))

    # Test seeded values.
    test_scene.setRngSeed(1234)
    orig_seeded_positions = test_scene.randomPositions()
    test_scene.setRngSeed(1234)  # reset seed
    new_seeded_positions = test_scene.randomPositions()
    assert np.all(np.equal(orig_seeded_positions, new_seeded_positions))


def test_collision_check(test_scene: Scene) -> None:
    # Collision free
    q_free = np.array([0.0, -1.57, 0.0, 0.0, 0.0, 0.0])
    assert not test_scene.hasCollisions(q_free)

    # In collision
    q_coll = np.array([0.0, -1.57, 3.0, 0.0, 0.0, 0.0])
    assert test_scene.hasCollisions(q_coll)


def test_collision_check_along_path(test_scene: Scene) -> None:
    # Collision free
    q_start_free = np.array([0.0, -1.57, 0.0, 0.0, 0.0, 0.0])
    q_end_free = np.array([1.0, -1.57, 1.57, 0.0, 0.0, 0.0])
    assert not hasCollisionsAlongPath(test_scene, q_start_free, q_end_free, 0.05)

    # In collision
    q_end_coll = np.array([0.0, -1.57, 3.0, 0.0, 0.0, 0.0])
    assert hasCollisionsAlongPath(test_scene, q_start_free, q_end_coll, 0.05)


def test_create_frame_map(test_scene: Scene) -> None:
    roboplan_examples_dir = Path(get_install_prefix()) / "share"
    roboplan_models_dir = roboplan_examples_dir / "roboplan_example_models" / "models"
    urdf_path = roboplan_models_dir / "ur_robot_model" / "ur5_gripper.urdf"
    package_paths = [roboplan_examples_dir]
    model, _, _ = pin.buildModelsFromUrdf(urdf_path, package_dirs=package_paths)
    for frame in model.frames:
        if frame.name == "universe":
            continue
        assert test_scene.getFrameId(frame.name) == model.getFrameId(frame.name)


def test_collision_geometry(test_scene: Scene) -> None:
    # Nominally, this configuration is collision free
    q = np.zeros(6)
    assert not test_scene.hasCollisions(q)

    # Add some collision objects
    color = np.array([0.5, 0.5, 0.5, 0.5])
    box_tform = np.eye(4)
    box_tform[2, 3] = 1.0  # z position
    test_scene.addBoxGeometry(
        "test_box", "universe", Box(1.0, 1.0, 0.5), box_tform, color
    )

    sphere_tform = np.eye(4)
    sphere_tform[0, 3] = 3.0  # x position
    test_scene.addSphereGeometry(
        "test_sphere", "universe", Sphere(0.5), sphere_tform, color
    )

    assert not test_scene.hasCollisions(q)  # should still be collision free

    # Now move one of the collision objects to be in collision.
    sphere_tform[0, 3] = 0.0
    test_scene.updateGeometryPlacement("test_sphere", "universe", sphere_tform)
    assert test_scene.hasCollisions(q)

    # Remove the collision object and verify that the robot is no longer in collision.
    test_scene.removeGeometry("test_sphere")
    assert not test_scene.hasCollisions(q)


def test_set_collisions(test_scene: Scene) -> None:
    # Nominally, this configuration is collision free
    q = np.zeros(6)
    assert not test_scene.hasCollisions(q)

    # Add a collision object such that the configuration is in collision.
    sphere_tform = np.eye(4)
    sphere_tform[0, 3] = 0.6
    test_scene.addSphereGeometry(
        "test_sphere",
        "universe",
        Sphere(0.1),
        sphere_tform,
        np.array([0.5, 0.5, 0.5, 0.5]),
    )
    assert test_scene.hasCollisions(q)

    # Use the frame names, which should automatically look up the corresponding collision geometries.
    test_scene.setCollisions("forearm_link", "test_sphere", False)
    assert not test_scene.hasCollisions(q)

    # Now re-add the collision pair, which should re-enable collision.
    test_scene.setCollisions("test_sphere", "forearm_link", True)
    assert test_scene.hasCollisions(q)

    # Add an invalid collision pair for check for errors.
    with pytest.raises(RuntimeError) as exc_info:
        test_scene.setCollisions("nonexistent_link", "test_sphere", True)
    expected_error = (
        "Could not set collisions: Could not get collision geometry IDs: "
        "Frame name 'nonexistent_link' not found in frame_map_."
    )
    assert str(exc_info.value) == expected_error
